dlib/cuda/cuda_dlib.h

```diff
@@ -439,7 +438,7 @@ namespace dlib
                 const size_t bytes_per_plane = subnetwork_output.nr()*subnetwork_output.nc()*sizeof(uint16_t);
                 // Allocate a cuda buffer to store all the truth images and also one float
                 // for the scalar loss output.
-                cuda_data_void_ptr buf = work->get(subnetwork_output.num_samples()*bytes_per_plane + sizeof(float));
```

Comment:
This can't go here, since once buf goes out of scope the buffer may be deallocated and if do_work() launched any cuda kernels that are still using it things will blow up.  It happens to be the case that do_work() doesn't currently do that, it syncs at the end, but who knows, maybe it will be different in the future.  And in any case, cuda allocations take forever, so we should be careful to avoid them. 

So instead, you could keep the class member and just check if it's initialized here and allocate it if it's not initialized. That would make everything work out properly I think.

Reply:
Ok. Did think about that, but figured (incorrectly it seems) that we would then need to have a separate work buffer for each thread / device (=> a `vector` or `unordered_map` of these `shared_ptr`s, or something like that... which would then need to be protected by a `mutex` as well, if new items could be added simultaneously by multiple threads).

But looking at it again, it seems that there are as many `compute_loss_multiclass_log_per_pixel` objects as there are threads / devices. So the above shouldn't be a problem.

So... just pushed a suggested fix to this problem.

Reply:
That’s essentially what device_global_buffer is, a thread and device specific buffer.  It’s held in a weak_ptr so it gets deallocated if no o e is using it in a specific device/thread combination. 

But right, any of the layer objects are only used in one device. 

This class should have a custom copy constructor too so that the work buffer never gets copied to another thread. Otherwise you could get a work buffer from the wrong thread which could be bad. The global_device_buffer method should probably return a unique_ptr to make this kind of mistake less likely now that I’m thinking about the use patterns more.

Reply:
Although I guess you can’t use unique_ptr with weak_ptr. Hmm. It would be cool if somehow copying the buffer pointers to other threads was difficult. Not sure how to accomplish that though.

Reply:
Need a non-copyable shared pointer 😅

Reply:
Anyway, don’t worry about this part in this PR. I’ll do something about it in a bit.